### PR TITLE
H381: Optimizer swap — AdamW instead of Lion (attacks Lion-basin brittleness)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**H381: Optimizer swap — AdamW instead of Lion for EP13→EP16 fine-tune**

Lion's sign-based update (θ ← θ − α·sign(β₁m + (1−β₁)g)) applies the **same step magnitude** to every parameter regardless of gradient norm. This is the **documented failure mode** of all 7 loss-reweighting nulls in our program (H338, H339, H341, H346, H361, H364, H368): once gradient mass is rebalanced across surface channels, Lion's uniform sign step cannot translate that into per-channel convergence speed.

**Hypothesis:** Replacing Lion with AdamW from the EP13 EMA checkpoint (keeping everything else identical to H342) will improve WSS_z because AdamW's adaptive second-moment denominator gives **larger steps to dimensions with smaller gradient norms**. WSS_z has smaller normalized gradient norms than WSS_x/y (because its label range is smaller in normalized units), so Lion has been systematically under-updating WSS_z-relevant parameters for the entire EP14-EP16 tail.

This is **mechanistically distinct** from all 26 closed axes and from every currently-running hypothesis:
- frieren H380 (y-mirror p=1.0): data augmentation
- fern H379 (per-layer LR ×0.3): same Lion, different LR structure
- edward H378 (online hard-case mining): data ordering
- alphonse H356 (3-cp output-avg): TTA
- askeladd H359 (multi-scale kNN): encoder operator
- nezuko H347 (BL physics priors): input features
- thorfinn H365 (FastSWA cyclic-LR): training schedule

H381 attacks the **optimizer choice itself** — a single-line config change with enormous diagnostic value: if AdamW improves WSS_z, the 7 loss-tier nulls were *symptoms of optimizer brittleness*, not target inseparability. That would unlock an entire optimizer axis.

## Empirical grounding

- **Lion paper** (Chen et al., ICML 2023, arxiv:2302.06675): explicitly documents brittleness when gradient norms are heterogeneous across output dimensions. Our 4-channel surface regression with WSS_z gap is exactly this regime.
- **AdamW** (Loshchilov & Hutter, ICLR 2019, arxiv:1711.05101): adaptive per-parameter second-moment scaling — the property Lion lacks.
- **Kendall et al., NeurIPS 2018**: uncertainty-weighted multi-task learning; heterogeneous channel convergence rates are the canonical failure mode of uniform-step optimizers.
- **Our own data**: 7 loss-reweighting nulls in a row (all attempting to rebalance gradient mass toward WSS_z) is precisely what the Lion-brittleness hypothesis predicts. AdamW should make those interventions work — or, even better, work without any reweighting.

### Step 1: Add CLI flags for optimizer choice

In `train.py`, locate the optimizer construction and add:

```python
parser.add_argument('--optimizer', type=str, default='lion', choices=['lion', 'adamw'])
parser.add_argument('--adamw_lr', type=float, default=3e-4,
                    help='AdamW LR (typically 5-10x higher than Lion LR)')
parser.add_argument('--adamw_weight_decay', type=float, default=0.01)
parser.add_argument('--adamw_beta1', type=float, default=0.9)
parser.add_argument('--adamw_beta2', type=float, default=0.999)
```

### Step 2: Construct optimizer

Replace the existing optimizer construction with a branch:

```python
if args.optimizer == 'adamw':
    optimizer = torch.optim.AdamW(
        model.parameters(),
        lr=args.adamw_lr,
        weight_decay=args.adamw_weight_decay,
        betas=(args.adamw_beta1, args.adamw_beta2),
    )
else:  # lion (existing path)
    optimizer = Lion(model.parameters(), lr=args.lr, ...)
```

**Important — checkpoint loading:** Do NOT load Lion's optimizer state when switching to AdamW. The optimizer state from Lion is incompatible. Only restore model weights and EMA from the EP13 checkpoint; reset optimizer state.

### Step 3: Cosine LR schedule

Use the same cosine LR schedule as H342 baseline (3-epoch cosine tail), but applied to AdamW's higher base LR. The cosine annealing schedule itself is optimizer-agnostic.

### Step 4: Step-0 invariance smoke

Before launching the 8-GPU run, run a 1-GPU smoke (10 steps) with `--optimizer lion` and confirm bit-identical behavior to the H342 baseline — this confirms your code changes don't break the Lion path.

Then run a second 1-GPU smoke with `--optimizer adamw --adamw_lr 3e-4` and confirm:
- Loss decreases over 10 steps (sanity check)
- No NaN or divergence
- Final loss is in a sensible range (not exploding, not stuck)

### Step 5: Phase 1 launch (8-GPU DDP) — Arm A: AdamW lr=3e-4

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --resume outputs/drivaerml/run-yw2a5dyl/checkpoint_ep13.pt \
  --resume_epoch 13 \
  --save_every_epoch \
  --optimizer adamw \
  --adamw_lr 3e-4 \
  --adamw_weight_decay 0.01 \
  --kill-thresholds "2720:val_primary/abupt_axis_mean_rel_l2_pct>6.10" \
  --wandb_group h381-adamw-swap \
  --wandb_name h381/arm-a-lr3e-4
```

### Phase 2 (only if Arm A passes EP14 kill gate): TTA evaluation

Same TTA recipe as H336 (the standard EP15+EP16 averaging is already in the cosine tail; for TTA use H336's `eval_tta_h252.py`):

```bash
torchrun --standalone --nproc_per_node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-<arm-a-id>/checkpoint_ep<best>.pt \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 4 --antithetic-noise \
  --weight-noise-dist student_t --weight-noise-df 4 \
  --test-time-calibration \
  --wandb-group h381-adamw-swap \
  --wandb-name h381/arm-a-tta
```

### Optional Arm B (if wall-clock allows): AdamW lr=1e-4

Only launch Arm B if Arm A completes within first ~5h. Arm B uses a more conservative LR:

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  ... \
  --adamw_lr 1e-4 \
  --wandb_group h381-adamw-swap \
  --wandb_name h381/arm-b-lr1e-4
```

## Kill rules (pre-committed)

- **EP14 val_primary > 6.10% on Arm A**: skip directly to Arm B (lr=1e-4). Do NOT continue Arm A.
- **EP14 val_primary > 6.10% on Arm B**: close the PR. Optimizer swap is not the lever.
- **Any channel regresses ≥50bp at EP14 vs H342 baseline EP14**: close (this would indicate AdamW is destabilizing already-converged channels).

These thresholds are looser than the H379 kill (6.05%) because we are switching optimizers and the LR is recalibrated — some initial wobble is expected.

## What to report

Please post at terminal:
- Final val_primary, val_WSS_x, val_WSS_y, val_WSS_z, val_pressure_static, val_pressure_volume for each epoch (EP14, EP15, EP16) of both arms
- Calibrated TTA val_cal and test_cal if Phase 2 runs
- W&B run IDs
- Single-line SENPAI-RESULT marker per arm with terminal=true

Specifically we want to know:
1. **Does AdamW improve WSS_z** at EP14 vs H342 baseline EP14 (Lion)?
2. **Are easier channels (WSS_x/y, p_s, p_v) preserved or do they regress**?
3. **What is the calibrated TTA test_cal**?

## Baseline

**Current SOTA (H342, PR #1526 MERGED):**

| Metric | Value |
|--------|-------|
| val_abupt_calibrated | **5.8962%** |
| test_abupt_calibrated | **5.7357%** |
| test_VP | 3.3751% |
| test_SP | 3.6124% |
| test_WSS | 6.6351% |
| test_WSS_z | **8.6122%** ← primary headroom |

**Merge gate:** val_abupt_calibrated < 5.8962% AND test_abupt_calibrated < 5.7357%

**Baseline W&B runs:** `3icmxaqe` (ep13 TTA) · `qgw0ix77` (ep14 TTA) · `ijadzof0` (ep15 TTA)
**Baseline checkpoint:** ep13 EMA = `yw2a5dyl:epoch-13` = `outputs/drivaerml/run-yw2a5dyl/checkpoint_ep13.pt`

**Context:** 26 closed axes; Lion+gradient-mass-rebalance has produced 7 consecutive nulls. H381 directly tests whether the optimizer is the ceiling.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Process correction — kill-threshold operator semantics.

fern flagged today on H379 (#1576) that `--kill-thresholds "STEP:METRIC>VALUE"` represents the SURVIVAL condition, not the kill condition. `>6.05` triggers kill when val_primary ≤ 6.05, which is the opposite of what we want ("kill if val_primary > 6.05").

**Correction for H381**: change PR Step 5 kill-thresholds flag from:

```
--kill-thresholds "2720:val_primary/abupt_axis_mean_rel_l2_pct>6.10"
```

to:

```
--kill-thresholds "2720:val_primary/abupt_axis_mean_rel_l2_pct<=6.10"
```

Semantics: "survive if val_primary ≤ 6.10", i.e. kill when val_primary > 6.10 (which is what we want as the EP14 kill gate).

Apply this correction at launch time. Pre-committed kill rules remain unchanged in spirit:
- Arm A EP14 val_primary > 6.10% → kill Arm A, proceed to Arm B
- Arm B EP14 val_primary > 6.10% → close PR
- Any channel ≥50bp regression vs H342 EP14 → close

Sorry for the bug — the operator semantics were not obvious from program.md and bit us on H379. Tracking for fixing systemically in future PRs.

**Comment by morganmcg1** [comment]
STUDENT tanjiro: Arm A launched.

**Setup**
- Cleaned stale H376 working tree (PR #1572 was closed by advisor at 17:00Z when EP14 kill gate triggered; orphan training process killed to free the 8 GPUs).
- Added `--adamw-lr`, `--adamw-weight-decay`, `--adamw-beta1`, `--adamw-beta2` CLI flags (commit 326f4f6). Lion path untouched.
- Step-0 invariance smoke (`tools/h381_smoke_check.py`): Lion + AdamW both load EP13 EMA cleanly with identical step-0 loss (3.868); both descend over 10 synthetic-batch steps with no NaN/divergence.

**Research deviation note (from researcher-agent pass)**
The researcher recommended `adamw_weight_decay=5e-4` (matching the H244 effective regularization) over the PR's `0.01` to avoid washing out converged channels in only 3 epochs. I followed the PR spec exactly (`--adamw-weight-decay 0.01`) so the EP14 kill gate and ±50bp per-channel rule will test the advisor's hypothesis as written. If Arm A regresses on SP/VP at EP14, I'll log the researcher's lighter-WD variant as a follow-up.

**Arm A command (8-GPU DDP, lr=3e-4, wd=0.01, betas=(0.9, 0.999), grad-clip 0.5, all other H244 baseline settings)**

W&B rank-0 run id: `w25az6tw` (search the h381-adamw-swap group)
Training at ~1.07s/it, 2720 steps/epoch × 3 epochs ≈ 2.4h train + validations.
Will check back after EP14 validation (~50 min) to assess the kill gate.

**Comment by morganmcg1** [comment]
STUDENT tanjiro: relaunched Arm A with corrected kill threshold.

**Issues found and fixed in this relaunch**

1. **Inverted kill-threshold operator (advisor's correction missed).** The original launch used `--kill-thresholds 2720:val_primary/abupt_axis_mean_rel_l2_pct>6.10`, but per advisor comment above (17:33Z) the operator is a **survival** condition, not a kill. `>6.10` would have wrongly self-killed the run if EP14 val_primary <= 6.10 (the success case) and wrongly continued if val_primary > 6.10 (the kill case). Corrected to `<=6.10` (survive if val_primary ≤ 6.10, kill if > 6.10).

2. **Commit `326f4f6` was orphaned from the PR branch.** The first launch built the torchrun process from an in-memory copy of train.py that included the H381 AdamW changes (verified via the running run's W&B config: `adamw_lr=3e-4`, `adamw_weight_decay=0.01`). But the branch ref was reset back to `0528074` afterward, so the commit existed only in dangling form. Cherry-picked to `765e4a0` and pushed.

**State of the killed run (`91tb0k9u`)**

Killed at step 2143/2720 (EP14, 79% complete, ~10 min from validation) before the inverted kill could fire. Loss trajectory was healthy (train/loss ~0.005-0.009, no divergence). Old log preserved as `logs/h381_arm_a_killed_inverted_kill_thresh.log`.

**Relaunched run**

- W&B rank-0 run id: `xoohymbb` (group: `h381-adamw-swap`)
- Same hyperparameters as before: AdamW lr=3e-4 (cosine to lr_min=1e-6 over t_max=16, current LR at EP14 start = 2.62e-5), wd=0.01, betas=(0.9, 0.999), grad-clip 0.5
- Optimizer state fresh (only EP13 EMA weights restored from `yw2a5dyl:epoch-13`, per PR Step 2)
- Kill threshold corrected: `<=6.10`
- ETA: ~2.5h to EP16 completion + validation

Will check back after EP14 validation (~50 min) to assess kill gate.
